[PYTX-VPDQ] Implement VPDQ faiss index

tigerchensh opened the pull request:
Summary
---------

Implement VPDQ FlatHashIndex which can return the percentage match in the indexMatch as well as matched features. When adding to the VPDQ Index, it requires metadata which includes the video_id: [str or int].

Test Plan
---------
A simple test wich runs through query_raw_result and query (returns match percentage).
Will add more tests in separate PR.
python3 -m pytest test_vpdq_faiss.py

tigerchensh commented on python-threatexchange/mypy.ini:
```diff
@@ -22,4 +22,7 @@ ignore_missing_imports = True
 [mypy-vpdq.*]
+ignore_missing_imports = True
+
+[mypy-faiss.*]
```
```
import faiss  # type: ignore
```
The import faiss in pdq_faiss_matcher.py is using type: ignore.

tigerchensh commented on python-threatexchange/threatexchange/extensions/video_vpdq/vpdq_faiss_matcher.py:
```diff
@@ -0,0 +1,183 @@
+        Returns:
+        sequence of matches per query
+            For each query provided in queries, the returned sequence will contain a sequence of matches within the index
+            that were within threshold hamming distance of that query. These matches will be (id, video_id, frame_number,
```
The id is the faiss_id when adding the signle vpdq feature. The video id represents a video which is a list of vpdq features

tigerchensh commented:
Thanks for the suggestion on the design! I have removed the video_id and made dedupe and quality filter before adding to the index. The vpdq_faiss.py now is more flexible with changing to different faiss index types.

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,137 @@
+        super().__init__()
+        self.index: VPDQHashIndex = self._get_empty_index()
+        self.entries: t.List = []
+        self.idx_to_vpdq: t.List[t.Tuple[int, vpdq.VpdqFeature]] = []
```
This name surprised me a little bit - my best guess for alternatives:
1. faiss_id_to_entry_id
2. index_idx_to_entry_idx
(or some variation)

Also, I'm not 100% you actually need the original VpdqFeature anywhere that you would need this unless you plan on verifying with the brute force. You might be doubling your storage cost without too much benefit, but I don't know if you are relying on this for testing.

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,137 @@
+            matches[hash] = match_tuples
+        return matches
+
+    def query_with_match_percentage_in_result(
```
Doesn't `query` return a VPDQMatchResult too with percentage in the result? Is there's something else that distinguishes these functions?

Dcallies commented on python-threatexchange/threatexchange/signal_type/index.py:
```diff
@@ -48,6 +48,19 @@ def __init__(self, distance: int, metadata: T) -> None:
         self.metadata = metadata
 
 
+class VPDQIndexMatch(IndexMatch):
+    query_match_percent: float
+    compared_match_percent: float
+
+    def __init__(
+        self, query_match_percent: float, compared_match_percent: float, metadata: T
+    ) -> None:
+        # Dist is meaning less here, because VPDQ match does not return a single dist
+        IndexMatch.__init__(self, -1, metadata)
+        self.query_match_percent = query_match_percent
+        self.compared_match_percent = compared_match_percent
+
+
```
blocking: This is a detail of vPDQ, which is not installed by default by the library, and so sohuld live in the vPDQ directory

Dcallies commented on python-threatexchange/threatexchange/signal_type/index.py:
```diff
@@ -48,6 +48,19 @@ def __init__(self, distance: int, metadata: T) -> None:
+        self, query_match_percent: float, compared_match_percent: float, metadata: T
+    ) -> None:
+        # Dist is meaning less here, because VPDQ match does not return a single dist
+        IndexMatch.__init__(self, -1, metadata)
```
blocking: You should never call the superclass like this, instead always use super().

The reason for this is that in multiple inheritence, or if someone refactors in a middle layer, this will break the expected behavior of initialization, and possibly be very hard to discover

Dcallies commented on python-threatexchange/threatexchange/signal_type/index.py:
```diff
@@ -48,6 +48,19 @@ def __init__(self, distance: int, metadata: T) -> None:
 
 
+class VPDQIndexMatch(IndexMatch):
+    query_match_percent: float
+    compared_match_percent: float
```
(not sure): You might be able to leave these out, and the typechecker can figure it out from reading the init, but I don't know

Dcallies reviewed (changes requested):
Hey Xiangnan, thanks for the continued work. This PR is basically the biggest chunk of functionality add, and maybe the pivotal PR of the integration, so I'm giving it a high level of scruitany. If you want to try and break it up or land some stuff incrementally so it doesn't feel like we are going back and forth on this one forever, let me know and I can help with that.

Toplevel thoughts:
* Don't use the 'bare' version of generics like t.List, t.Dict, etc. These implicitly turn into t.List[t.Any]. If you really mean Any, use that specifically, but usually you know the type. Each t.Any or #type: ignore weakens the benefits of typechecking. There's a lint rule we can add to warn on this.
* Consider helping a reader figure out which methods are an internal detail, and which ones are part of the public interface. Prefixing methods with `_` like `_private_method()` can help a reader understand that this is an implementation detail that they should not rely on externally.

Originally I wrote you something long here about the need for deduping hashes from the index first. However, it looks like faiss still works without it, but it will become a performance problem in the future: https://github.com/facebookresearch/faiss/wiki/FAQ#searching-duplicate-vectors-is-slow

Up to you if you want to try and fix it now (like I did in the image library I linked you) or just make a # todo for it

tigerchensh commented:
Thanks for the reviews and suggestions on the design. I have removed query_raw_result and moved query_percentage into query, since we only need percentages from VPDQIndexMatch.  I have implemented supports for duplicated hashes in VPDQIndex. Write tests for adding duplicating hashes and serialization. Will add more tests in next PR.

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/tests/test_vpdq_faiss.py:
```diff
@@ -0,0 +1,71 @@
+        VPDQIndexMatch,
+    )
+
+ROOTDIR = Path(__file__).parents[5]
```
Is this still used?

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/tests/test_vpdq_faiss.py:
```diff
@@ -0,0 +1,71 @@
+
+@pytest.mark.skipif(_DISABLED, reason="vpdq not installed")
+def compare_match_result(res1, res2) -> bool:
+    return (
+        res1.compared_match_percent == res2.compared_match_percent
+        and res1.query_match_percent == res2.query_match_percent
+        and res1.metadata == res2.metadata
+        and res1.distance == res2.distance
+    )
```
(ignorable): You can also implement `__eq__` on the object, which will let you do `==` directly, you can look up how to do that (there some tricks)

tigerchensh replied:
I don't think there is any use case for comparing two IndexMatch results except for testing. So, I think __eq__ is overkill.

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/tests/test_vpdq_faiss.py:
```diff
@@ -0,0 +1,71 @@
+example_meta_data = {"name": "example_video"}
+
+
+@pytest.mark.skipif(_DISABLED, reason="vpdq not installed")
```
Check this out on how to skip all the tests at once: https://docs.pytest.org/en/7.1.x/how-to/skipping.html#skip-all-test-functions-of-a-class-or-module

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_faiss.py:
```diff
@@ -0,0 +1,86 @@
+            "0000000000000000000000000000000000000000000000000000000000000000" for a threshold of 16. Thus it would appear in
+            the entry for both the hashes if they were both in the queries list.
+
+            eg.
```
nit: e.g.

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_faiss.py:
```diff
@@ -0,0 +1,86 @@
+            result[query.hex] = list(zip(matches, distances))
+        return result
+
+    def __getstate__(self):
+        data = faiss.serialize_index_binary(self.faiss_index)
+        return data
+
+    def __setstate__(self, data):
+        self.faiss_index = faiss.deserialize_index_binary(data)
```
👍

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
+Implementation of SignalTypeIndex abstraction for PDQ by wrapping
```
vPDQ

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+
+"""
+Implementation of SignalTypeIndex abstraction for PDQ by wrapping
+hashing.pdq_faiss_matcher.
```
nit: update

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+
+    def __init__(
+        self,
+        quality_threshold: int = VPDQ_QUALITY_THRESHOLD,
```
👍

(ignorable): You may want to consider forcing this to be a kwarg by insert `*,` after self

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+        if unique_features:
+            self.index.add_single_video(unique_features)
+
+    def query(self, query_hash: str) -> t.List[IndexMatch[IndexT]]:
```
Does the typer let you use VPDQIndexMatch here?

tigerchensh replied:
No, mypy doesn't think return type VPDQIndexMatch matches IndexMatch[Any]

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+                    if entry_id not in index_matched:
+                        index_matched[entry_id] = set()
+                    index_matched[entry_id].add(vpdq_match)
+                # Dist(-1) is meaningless here, because VPDQ match does not return a single dist
```
Misplaced comment

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+        for entry_id in query_matched.keys():
+            matches.append(
+                VPDQIndexMatch(
+                    -1,
```
Suggest using the match percentage (int, 0-100) as the placeholder here, instead of -1, which is a bit misleading.

I agree with you that the types are mismatched here. I would like to redo this to make the distance type an internal detail

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+            matches.append(
+                VPDQIndexMatch(
+                    -1,
+                    len(query_matched[entry_id]) * 100 / len(features),
+                    len(index_matched[entry_id])
+                    * 100
```
Super dumb question - should we consider using 0-1.0 instead of 0-100?

tigerchensh replied:
What is the benefit of using 0-1.0 instead of 0-100? I feel like 0-100 has more readability and it can still have the precision as a float.

Dcallies replied:
When you are multiplying things together like durations you don't have to re-normalize it to a fraction again.

```
matching_sec = video_duration * match_pct
```
vs
```
matching_sec = video_duration * match_pct/100
```

The only time when 0-100 is preferred is for humans to look at., and that is what `__str__` is generally used for.

tigerchensh replied:
Yes, it is worth doing. But I will create a separate PR for making all the matching percentage to be 0-1.0

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+            matches.append(
+                VPDQIndexMatch(
+                    -1,
+                    len(query_matched[entry_id]) * 100 / len(features),
+                    len(index_matched[entry_id])
+                    * 100
+                    / len(self._entry_idx_to_features_and_entires[entry_id][0]),
```
This is slightly hard to read, might be worth splitting out the variables here so it's clear what each one means.

tigerchensh replied:
Thanks for pointing out!

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_util.py:
```diff
@@ -11,6 +11,10 @@
 TIMESTAMP = "timestamp"
 VPDQ_TIMESTAMP_PRECISION = 3
+BITS_IN_VPDQ = 256
+VPDQ_VIDEOID_TYPE = t.Union[str, int]
```
Is this still used?

tigerchensh replied:
Not anymore, removed

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_util.py:
```diff
@@ -11,6 +11,10 @@
 HASH = "hash"
 TIMESTAMP = "timestamp"
 VPDQ_TIMESTAMP_PRECISION = 3
+BITS_IN_VPDQ = 256
```
(ignorable): I think you could use BITS_IN_PDQ since this is part of the threatexchange library

Dcallies commented on python-threatexchange/threatexchange/signal_type/index.py:
```diff
@@ -48,6 +48,19 @@ def __init__(self, distance: int, metadata: T) -> None:
         self.metadata = metadata
 
 
+class VPDQIndexMatch(IndexMatch):
```
You have to pass through T here to get it type 
```suggestion
class VPDQIndexMatch(IndexMatch[T]):
```

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/tests/test_vpdq_faiss.py:
```diff
@@ -0,0 +1,71 @@
+    )
+
+ROOTDIR = Path(__file__).parents[5]
+example_meta_data = {"name": "example_video"}
```
Traditionally these are in ALL_CAPS

You can also type this as t.Mapping[str, str]  to enforce that it can't accidentally be changed.

You could also add `hash` here as well

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/tests/test_vpdq_faiss.py:
```diff
@@ -0,0 +1,71 @@
+    assert len(index._index_idx_to_vpdqHex_and_entry) == len(features)
+    res = index.query(hash)
+    # A complete match to itself
+    assert compare_match_result(res[0], VPDQIndexMatch(-1, 100, 100, example_meta_data))
+
+
+@pytest.mark.skipif(_DISABLED, reason="vpdq not installed")
+def test_serialize():
+    hash = VideoVPDQSignal.get_examples()[0]
+    index = VPDQIndex.build([[hash, example_meta_data]])
+    pickled_data = pickle.dumps(index)
+    reconstructed_index = pickle.loads(pickled_data)
+    res = reconstructed_index.query(hash)
+    assert compare_match_result(res[0], VPDQIndexMatch(-1, 100, 100, example_meta_data))
+
+
+@pytest.mark.skipif(_DISABLED, reason="vpdq not installed")
+def test_duplicate_hashes():
+    hash = VideoVPDQSignal.get_examples()[0]
+    index = VPDQIndex.build([[hash, "video1"]])
+    index.add(hash, "video2")
+    res = index.query(hash)
+    # A complete match to itself
+    assert compare_match_result(res[0], VPDQIndexMatch(-1, 100, 100, "video1"))
+    assert compare_match_result(res[1], VPDQIndexMatch(-1, 100, 100, "video2"))
```
All your tests are all asserting that they are perfect matches, which means that many failure modes might be missed. I am not yet convinced that your implementation is correct as a result. Would like to see some variation here in a followup PR.

i.e.
```
pdq_dist(A, A) = 0
pdq_dist(A, B) = 31
pdq_dist(A, C-F) > 31

v1 = [A]
v2 = [A, B]
v3 = [B]
V4 = [B, C]
V5 = [B, C, D]

for x in range(1, 6):
  for y in range(x, 6):
    vpdq_dist(v{x}, v{y})
    vpdq_dist(V{y}, V{x})

```

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+            t.Tuple[t.List[vpdq.VpdqFeature], IndexT]
+        ] = []
+        self._index_idx_to_vpdqHex_and_entry: t.List[t.Tuple[int, t.List[int]]] = []
+        self._unique_vpdqHex_to_idx: t.Dict[str, int] = {}
```
Which idx? We have index_idx and entry_idx (assume entry)

tigerchensh replied:
It's actually  _unique_vpdqHex_to_index_idx, because each unique hex hash should only map to one index_idx in faiss.

Dcallies commented on python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py:
```diff
@@ -0,0 +1,100 @@
+        entry_id = len(self._entry_idx_to_features_and_entires)
+        features = prepare_vpdq_feature(signal_str, self.quality_threshold)
+        self._entry_idx_to_features_and_entires.append((features, entry))
+        # Use hex to represent the feature because it saves the space
```
There are even more compact serialization, such as bytes.

tigerchensh replied:
I think I will leave it as it is for easier checking if the feature(hex) already exists in _unique_vpdqHex_to_idx.

Dcallies reviewed (commented):
Didn't finish, flushing

Dcallies reviewed (approved):
I have no more blocking comments, so accepting.

However, please sync with me before proceeding to the next block of work - I think we should invest significantly in testing in verification, rather than new features. Can talk to you about what I mean offline.

tigerchensh commented:
Thanks for the reviews and suggestions. Will move percetange change (0-100 to 0-1.0) and more through testing in separate prs.
